File: backend-python/pcm_utils.py
import re
from difflib import SequenceMatcher

def clean_text(text: str) -> str:
    """Remove pontuação e converte para minúsculas."""
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)
    return text

def calculate_pcm(original_text: str, transcribed_text: str) -> dict:
    """
    Calcula o PCM comparando o texto original com o transcrito.
    Assume que a gravação teve exatamente 60 segundos.
    """
    orig_clean = clean_text(original_text).split()
    tran_clean = clean_text(transcribed_text).split()

    matcher = SequenceMatcher(None, orig_clean, tran_clean)
    matches = matcher.get_matching_blocks()
    
    total_correct = sum(match.size for match in matches)
    
    # Identificação simples de erros (baseado no que sobrou)
    # Futuramente, a IA fará uma classificação mais profunda
    diffs = {
        "corretas": total_correct,
        "total_original": len(orig_clean),
        "total_lido": len(tran_clean),
        "precisao": round((total_correct / len(orig_clean)) * 100, 2) if orig_clean else 0
    }
    
    return diffs

def get_performance_level(pcm: int) -> str:
    """Classifica o nível com base no PCM conforme PRD."""
    if pcm <= 60:
        return "Fase Inicial"
    elif pcm <= 75:
        return "Em Desenvolvimento"
    elif pcm <= 95:
        return "Em Consolidação"
    else:
        return "Fluente"

File: backend/pcmUtils.js
const cleanText = (text) => {
    if (!text) return "";
    return text
        .toLowerCase()
        .normalize("NFD")
        .replace(/[\u0300-\u036f]/g, "") // Remove acentos
        .replace(/[^\w\s]/g, "") // Remove pontuação
        .replace(/\s+/g, " ") // Normaliza espaços
        .trim();
};

const calculatePCM = (originalText, transcribedText) => {
    const origClean = cleanText(originalText).split(" ");
    const tranClean = cleanText(transcribedText).split(" ");

    let totalCorrect = 0;
    let currentTranIdx = 0;

    // Algoritmo de busca sequencial para contar palavras corretas na ordem
    for (let i = 0; i < origClean.length; i++) {
        const targetWord = origClean[i];

        // Procura a palavra no que resta do texto transcrito
        // Limita a busca a uma janela razoável para evitar saltos irreais
        const searchRange = tranClean.slice(currentTranIdx, currentTranIdx + 5);
        const foundLocalIdx = searchRange.indexOf(targetWord);

        if (foundLocalIdx !== -1) {
            totalCorrect++;
            currentTranIdx += foundLocalIdx + 1;
        }
    }

    return {
        corretas: totalCorrect,
        total_original: origClean.length,
        total_lido: tranClean.length,
        precisao: origClean.length > 0 ? Number(((totalCorrect / origClean.length) * 100).toFixed(2)) : 0
    };
};

const getPerformanceLevel = (pcm) => {
    if (pcm <= 60) return "Fase Inicial";
    if (pcm <= 75) return "Em Desenvolvimento";
    if (pcm <= 95) return "Em Consolidação";
    return "Fluente";
};

module.exports = { calculatePCM, getPerformanceLevel };
